lib/puppet/configurer.rb

```diff
@@ -372,6 +372,14 @@ def run_internal(options)
       execute_postrun_command or return nil
     end
   ensure
+    if Puppet[:resubmit_facts]
+      # TODO: Should mark the report as "failed" if an error occurs and
+      #       resubmit_facts returns false. There is currently no API for this.
```

Comment:
If fact resubmission fails, then I'd expect the overall report to be marked as failed. The call to `finalize_report` calls `compute_status` and changes the report status from the default `failed` to `success`. You could change `attr_reader :status` to an accessor, but then it's harder to control the status. Perhaps better to modify `finalize_report` to accept an optional argument for cases where the report status cannot be determined based on the transaction completing, resource statuses, etc.:

```ruby
def finalize_report(status = nil)
  ...
  @status = status || compute_status(resource_metrics, change_metric)
```

Reply:
Updating `finalize_report` seems like it would only work in this specific circumstance where `resubmit_facts` is called a few lines before `finalize_report` and thus a value could easily be threaded between the two.

Having something like a `mark_failed` function that causes `compute_status` to return `failed` when called seems like it would work better without coupling things tightly.

However, given I'd like to land this in 5.5.x for current PE LTS users, I don't think mucking with the report API is on the table for this PR.

Reply:
`finalize_report` is the method where the status is computed, so I'd prefer not creating another method for marking the status (as then there are 2 ways to do something). Adding an optional parameter isn't a breaking change.

Reply:
But then any time something happens that should fail the report, it has to be threaded back to the `finalize_report` call. That doesn't seem tenable for things that don't happen to be located inside the `ensure` block.

Reply:
The problem with a stand alone `mark_failed` is we don't want the report to be marked as failed, but have no information about why it failed. For example https://tickets.puppetlabs.com/browse/PUP-8908. So instead the report success/failure is computed based on resource_statuses. And the status *defaults* to "failed", as there can be failures before we even get to the point where we can apply a catalog. So `finalize_report` is the place to decide whether we completed everything successfully, and only then change the overall status from failed to success.

Reply:
I'd also be fine adding a `marked_failed` method so long as it required a string or exception about why it failed, and logged the exception. Then the `resubmit_facts` method could rescue the exception and do:

```ruby
def resubmit_facts(report)
  ...
rescue => detail
   report.mark_failed(detail)
   return false
end
```

Reply:
Seems fine to me. However, it's still a broader scope than I'd like to tackle in this PR. Would it be fine to ticket that as a follow-up improvement?

Reply:
I know what you're saying... but I think handling possible failures introduced by this PR should be part of this PR and the TODO removed.

Reply:
@joshcooper Another ping on this. It will be a while before I can circle back and tackle something as big as the reporting change. However, I feel like the feature as-is represents a meaningful improvement.

Is there a reason we can't ship it now and iterate on the reporting?